File: python/rosalind_cons.py
from typing import Dict, List

def read_data(filepath: str) -> Dict[str, str]:
    with open(filepath, mode='r') as file:
        fasta_dict = {}
        lines = file.read().strip().split('>')

        for line in lines:
            
            if not line:
                continue

            elements = line.split()
            fasta_dict[elements[0]] = ''.join(elements[1:])
        
        return fasta_dict
        
def write_data(filepath: str, profile: Dict[str, List[int]], consensus: str) -> None:
    with open(filepath, mode='w') as file:
        file.write(consensus + '\n')
    
        for symbol in sorted(profile.keys()):
            file.write('{}: {}\n'.format(symbol, ' '.join(map(str, profile[symbol]))))

def get_profile(fasta_dict: Dict[str, str]) -> Dict[str, List[int]]:
    profile = {}

    for label, string in fasta_dict.items():
        for index, symbol in enumerate(string):

            if symbol not in profile.keys():
                profile[symbol] = [0] * len(string)

            profile[symbol][index] += 1

    return profile

def get_consensus(profile: Dict[str, List[int]]) -> str:
    consensus = ""
    
    profile_keys = list(profile.keys())
    consensus_length = len(profile[profile_keys[0]])
    
    for index in range(consensus_length):
        most_common_symbol = ('', 0)
        
        for symbol in profile_keys:
            current_symbol = (symbol, profile[symbol][index])
            
            if current_symbol[1] > most_common_symbol[1]:
                most_common_symbol = current_symbol
        
        consensus += most_common_symbol[0]
    
    return consensus

if __name__ == '__main__':
    data = read_data('input/rosalind_cons.txt')
    profile = get_profile(data)
    consensus = get_consensus(profile)
    write_data('output/rosalind_cons.txt', profile, consensus)     

File: java/RosalindCONS.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.*;
import java.util.stream.Collectors;

public class RosalindCONS {
    private static HashMap<String, String> readData(String filePath) throws IOException {
        Path inputFile = Paths.get(filePath);
        List<String> inputLines = Files.readAllLines(inputFile);

        HashMap<String, String> fastaMap = new HashMap<>();
        String label = "";

        for (String line : inputLines) {
            if (line.charAt(0) == '>') {
                label = line.substring(1);
                fastaMap.put(label, "");
                continue;
            }

            fastaMap.put(label, fastaMap.get(label) + line);
        }

        return fastaMap;
    }

    private static void writeData(String filePath, TreeMap<Character, List<Integer>> profile, String consensus) throws IOException {
        Path outputFile = Paths.get(filePath);

        Files.deleteIfExists(outputFile);
        Files.createDirectories(outputFile.getParent());
        Files.createFile(outputFile);

        Files.writeString(outputFile, consensus + "\n", StandardOpenOption.APPEND);

        for (Character key : profile.keySet()) {
            String joinedValues = profile.get(key).stream().map(String::valueOf).collect(Collectors.joining(" "));
            Files.writeString(outputFile, key + ": " + joinedValues + "\n", StandardOpenOption.APPEND);
        }
    }

    private static TreeMap<Character, List<Integer>> getProfile(HashMap<String,String> fastaMap) {
        TreeMap<Character, List<Integer>> profile = new TreeMap<>();

        for (String label : fastaMap.keySet()) {
            String dna = fastaMap.get(label);

            for (int i = 0; i < dna.length(); i++) {
                if (!profile.containsKey(dna.charAt(i))) {
                    List<Integer> values = new ArrayList<>(Collections.nCopies(dna.length(), 0));
                    profile.put(dna.charAt(i), values);
                }

                List<Integer> values = profile.get(dna.charAt(i));
                values.set(i, profile.get(dna.charAt(i)).get(i) + 1);
                profile.put(dna.charAt(i), values);
            }
        }

        return profile;
    }

    private static String getConsensus(TreeMap<Character, List<Integer>> profile) {
        List<Character> keys = new ArrayList<>(profile.keySet());

        String consensus = "";
        int consensusLength = profile.get(keys.get(0)).size();

        for (int i = 0; i < consensusLength; i++) {

            char mostCommonKey = 0;
            int mostKeyOccurrences = 0;

            for (Character key : keys) {
                int currentKeyOccurrences = profile.get(key).get(i);

                if (currentKeyOccurrences > mostKeyOccurrences) {
                    mostCommonKey = key;
                    mostKeyOccurrences = currentKeyOccurrences;
                }
            }

            consensus += mostCommonKey;
        }

        return consensus;
    }

    public static void main(String[] args) throws IOException {
        HashMap<String, String> data = readData("input/rosalind_cons.txt");
        TreeMap<Character, List<Integer>> profile = getProfile(data);
        String consensus = getConsensus(profile);
        writeData("output/rosalind_cons.txt", profile, consensus);
    }
}
